### PR TITLE
[proof] [sat] Avoid duplicating assumption in chain links for SAT proof

### PR DESCRIPTION
**Pull request by HanielB**
Previously the generation of proofs in the SAT proof manager was generating different ASSUMPTION proof nodes for the same assumptions. This led to duplication in the final SAT proof of the proofs of these respective nodes. When these nodes are justified via something whose post-processing is expensive, such as MACRO_RESOLUTION, we had performance issues. 

This commit addresses the issue by using a specialized buffered proof generator for the links that when chained compose the SAT proof. It caches and reuses ASSUMPTION proof nodes when generating the proof nodes for the respective conclusion of the links. In our regressions this change leads to a 17% reduction in total size of proofs.

This commit also fixes an issue in the ProofCnfStream where we were adding symmetry steps automatically, which could be problematic, as explained in the added documentation. This issue happens in our regression `regress1/strings/issue6184-unsat-core.smt2` after the changes for size reduction, and is (thus indirectly) duplicated by https://github.com/cvc5/cvc5-projects/issues/529. Likewise, we were not properly accounting for disabled symmetry in the LazyCDProof when asking for a generator for a fact. This commit also fixes this.

### CODE REVIEW

File: src/proof/buffered_proof_generator.h [hunk 1]
```diff
@@ -53,7 +51,9 @@ class BufferedProofGenerator : protected EnvObj, public ProofGenerator
   /** identify */
   std::string identify() const override { return "BufferedProofGenerator"; }
 
- private:
+ protected:
+  typedef context::CDHashMap<Node, std::shared_ptr<ProofStep>> NodeProofStepMap;
```
Comment (ajreynol) [inline src/proof/buffered_proof_generator.h:55]:
```suggestion
  using NodeProofStepMap = context::CDHashMap<Node, std::shared_ptr<ProofStep>>;
```

File: src/prop/sat_proof_manager.h [hunk 2]
```diff
@@ -40,6 +40,45 @@ namespace prop {
...
 
+/**
+ * Works exactly as the regular buffered proof generator but when retrieving the
+ * proof of a step, we reuse the same ASSUMPTION step for repeated
+ * premises. This can be done because in the SAT proof there are no SCOPE steps,
+ * so there is no danger of mixing the scopes of assumptions.
+ */
+class SatBufferedProofGenerator : public BufferedProofGenerator
+{
+  typedef context::CDHashMap<Node, std::shared_ptr<ProofNode>> NodeProofNodeMap;
```
Comment (ajreynol) [inline src/prop/sat_proof_manager.h:51]:
```suggestion
  using NodeProofNodeMap = context::CDHashMap<Node, std::shared_ptr<ProofNode>>;
```

File: src/prop/sat_proof_manager.cpp [hunk 3]
```diff
@@ -24,6 +24,47 @@
 namespace cvc5::internal {
 namespace prop {
 
+void SatBufferedProofGenerator::addStep(Node fact, ProofStep ps)
+{
+  Trace("test") << "??\n";
```
Comment (ajreynol) [inline src/prop/sat_proof_manager.cpp:29]: Delete?

**Review by ajreynol — CHANGES_REQUESTED** [review]
A couple minor comments.

File: src/prop/sat_proof_manager.cpp (hunk 3 above)
Comment (HanielB) [reply src/prop/sat_proof_manager.cpp]: Yes. :) Done.

File: src/prop/sat_proof_manager.h [hunk 4]
```diff
@@ -40,6 +40,45 @@ namespace prop {
 
 class CnfStream;
 
+/**
+ * Works exactly as the regular buffered proof generator but when retrieving the
+ * proof of a step, we reuse the same ASSUMPTION step for repeated
+ * premises. This can be done because in the SAT proof there are no SCOPE steps,
+ * so there is no danger of mixing the scopes of assumptions.
+ */
+class SatBufferedProofGenerator : public BufferedProofGenerator
```
Comment (ajreynol) [inline src/prop/sat_proof_manager.h:49]: Can we make this a flag on BufferedProofGenerator, like `mkUniqueAssume`? Similar to `autoSym` for CDProof.

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/prop/sat_proof_manager.h (hunk 4 above)
Comment (HanielB) [reply src/prop/sat_proof_manager.h]: Done

**Review by ajreynol — APPROVED** [review]
LGTM, thanks for the update.

**Pull request merged**